internal/engine/compiler/impl_amd64.go

```diff
@@ -3598,21 +3598,95 @@ func (c *amd64Compiler) compileLoadDataInstanceAddress(dataIndex uint32, dst asm
+	emptyEightGroupsJump := c.assembler.CompileJump(amd64.JEQ)
+
+	rvalues := []*runtimeValueLocation{destinationOffset, sourceOffset, copySize}
+
+	// Prepare registers for swaps. There will never be more than 3 XCHGs in total.
+	swaps := c.compilePreventCrossedTargetRegisters(rvalues, []asm.Register{amd64.RegDI, amd64.RegSI, amd64.RegCX})
+
+	// Prepare registers for REP MOVSQ: copy from rsi to rdi, rcx times.
+	c.compileMaybeSwapRegisters(destinationOffset.register, amd64.RegDI)
+	c.compileMaybeSwapRegisters(sourceOffset.register, amd64.RegSI)
+	c.compileMaybeSwapRegisters(copySize.register, amd64.RegCX)
+
+	// Point on first byte of first quadword to copy.
+	if backwards {
+		c.assembler.CompileConstToRegister(amd64.ADDQ, -int64(bwOffset), amd64.RegDI)
+		c.assembler.CompileConstToRegister(amd64.ADDQ, -int64(bwOffset), amd64.RegSI)
+		// rep direction backwards
+		c.assembler.CompileStandAlone(amd64.STD)
+	}
+
+	c.assembler.CompileStandAlone(amd64.REPMOVSQ)
+
+	if backwards {
+		// Reset direction.
+		c.assembler.CompileStandAlone(amd64.CLD)
+	}
+
+	// Restore registers.
+	c.compileMaybeSwapRegisters(destinationOffset.register, amd64.RegDI)
+	c.compileMaybeSwapRegisters(sourceOffset.register, amd64.RegSI)
+	c.compileMaybeSwapRegisters(copySize.register, amd64.RegCX)
+	c.compileRestorePreventedCrossing(rvalues, swaps)
+
+	c.assembler.SetJumpTargetOnNext(emptyEightGroupsJump)
+	c.assembler.CompileStandAlone(amd64.NOP)
 }
 
-// compileCopyImpl implements compileTableCopy and compileMemoryCopy.
-//
-// TODO: the compiled code in this function should be reused and compile at once as
-// the code is independent of any module.
-func (c *amd64Compiler) compileCopyImpl(isTable bool, srcTableIndex, dstTableIndex uint32) error {
-	outOfBoundsErrorStatus := nativeCallStatusCodeMemoryOutOfBounds
-	if isTable {
-		outOfBoundsErrorStatus = nativeCallStatusCodeInvalidTableAccess
+// compileMemoryCopyLoopImpl is used for directly copying after bounds/direction check.
+func (c *amd64Compiler) compileMemoryCopyLoopImpl(destinationOffset, sourceOffset, copySize *runtimeValueLocation, tmp asm.Register, backwards bool) {
+	// Point on first byte to be copied depending on direction.
+	if backwards {
+		c.assembler.CompileNoneToRegister(amd64.DECQ, sourceOffset.register)
+		c.assembler.CompileNoneToRegister(amd64.DECQ, destinationOffset.register)
+	} else {
+		c.assembler.CompileRegisterToRegister(amd64.SUBQ, copySize.register, sourceOffset.register)
+		c.assembler.CompileRegisterToRegister(amd64.SUBQ, copySize.register, destinationOffset.register)
 	}
 
+	// destinationOffset += memory buffer's absolute address.
+	c.assembler.CompileRegisterToRegister(amd64.ADDQ, amd64ReservedRegisterForMemory, destinationOffset.register)
+	// sourceOffset += memory buffer's absolute address.
+	c.assembler.CompileRegisterToRegister(amd64.ADDQ, amd64ReservedRegisterForMemory, sourceOffset.register)
+
+	beginLoop := c.assembler.CompileStandAlone(amd64.NOP)
```

Comment:
Comment what happens inside this loop like `Copy the remainder bytes divided by 8 so that we can copy the content in the 8 byte loop below`